### DISCUSSION TITLE
Adding more components to the target scoring function

### DISCUSSION THREAD

**Discussion by palazzef** [GT4SD/gt4sd-core#181] [Q&A]
I'd like to add more components to the target scoring function. In one of the notebook I found this nice example:

# maximizing drug likeness and synthesizability
target = {"qed": {"weight": 1.0},
          "sa": {"weight": 1.0}}

Which is the correct syntax to add ECFP2 tanimoto similarity? Or more in general, where do I have to check to see how to add components? Thanks for the help!

**Comment by jannisborn** [comment]
Hi Ferri,

thanks for your interest, here are the supported minimzation functions: https://github.com/GT4SD/gt4sd-core/blob/d6dd5919440cc6c5c891f9a7a9981c27fc63ece6/src/gt4sd/algorithms/controlled_sampling/paccmann_gp/implementation.py#L51

It does not contain ECFP2 similarity. But this model relies on the `paccmann_gp` package which has a callable minimization function that supports arbitrary callables. See here: https://github.com/PaccMann/paccmann_gp/blob/main/paccmann_gp/callable_minimization.py

You would have to define the function yourself and then pass this just like the others. 
To work properly, this will require a change in the gt4sd code. In the line I linked above, it might suffice to add `'callable': 'CallableMinimization'`

**Reply by jannisborn** [reply]
If this line suffices or you find another way to make this work, we would be delighted to see a PR and integrate this into the package for future releases.

**Reply by jannisborn** [reply]
@palazzef this feature will be released in the next version, see the PR I opened #182

**Reply by palazzef** [reply]
Great, thanks!

**Comment by palazzef** [comment]
Hi @jannisborn, maybe I have to expand a bit more my question.

Basically I would like to expand the conditional part of the discovery-demo example by adding more components (qed+similarity+sa) and also more generative models (AAE, VAE, ORGAN etc..).

For this reason, I am looking for a mix&match version of "target=gentrl_ddr1_smi" (similarity only) and "target={"qed": {"weight": 1.0},
"sa": {"weight": 1.0}}" (QED and SA only).

**Comment by jannisborn** [comment]
Adding more models should be easy, I recommend checking the docs, e.g., see the GuacaMol models; https://gt4sd.github.io/gt4sd-core/api/gt4sd.algorithms.conditional_generation.guacamol.html

There's also a function called `gt4sd.algorithms.registry.list_available` that shows all algorithms including versions.

For the mix & matching: If you want to plot the SA instead of QED, can use the SA property predictor:
```console
>>>from gt4sd.properties import PropertyPredictorRegistry
>>>sa = PropertyPredictorRegistry.get_property_predictor("sas")
>>>sa('CCO')
```

If you want to have a generative model that simultaneously optimizes similarity to a seed molecule, QED and SA, then, AFAIK, PaccMannGP is the only model that supports this and the way to achieve this is the one I described above. Hope this helps!

**Comment by palazzef** [comment]
Thanks for the answer.

I thought that the models with "get_target_description()" function (like SMILESGAGenerator) were able to do MPO given a specific multiparameter target function.

**Reply by jannisborn** [reply]
For every model, that function just gives the description of the target value. But you're right for some of the Guacamol models, it should be possible to do such MPO

**Reply by palazzef** [reply]
ok, so if I can suggest a possible change (related also to #182), I will add a general "scoring function builder" that people can then easily use within several models (some of the guacamol models, reinvent, pacmannGP, and so on..). Maybe it is already like this, sorry I haven't had time to check the code in details.

**Reply by jannisborn** [reply]
This is available in a base form under `gt4sd.properties.scores`, see here: https://github.com/GT4SD/gt4sd-core/blob/main/src/gt4sd/properties/scores/core.py.

I'm not 100% sure that you can use the exakt same syntax for all models (GuacaMool, Reinvent, PaccMannGP), maybe @drugilsberg knows this from the top of his head.

There is even a `PropertyPredictorScorer` class that should allow to use all the property predictors from `gt4sd.properties` as scoring functions. See also: https://github.com/GT4SD/gt4sd-core/pull/124
